lib/actions/deploy.js

```diff
@@ -54,110 +54,113 @@ const deployProject = function(docker, logger, composeOpts, opts) {
-						);
-						return {};
+			return (
+				Promise.map(project.descriptors, function(d) {
+					// unconditionally build (or pull) if explicitly requested
+					if (opts.shouldPerformBuild) {
+						return d;
 					}
-					return compose
-						.buildProject(
-							docker,
-							logger,
-							project.path,
-							project.name,
-							compositionToBuild,
-							opts.app.arch,
-							opts.app.device_type,
-							opts.buildEmulated,
-							opts.buildOpts,
-							composeOpts.inlineLogs,
-							opts.convertEol,
-							composeOpts.dockerfilePath,
-							composeOpts.nogitignore,
-						)
-						.then(builtImages => _.keyBy(builtImages, 'serviceName'));
+					return docker
+						.getImage(typeof d.image === 'string' ? d.image : d.image.tag)
+						.inspect()
+						.return(d.serviceName)
+						.catchReturn();
 				})
-				.then(builtImages =>
-					project.descriptors.map(
-						d =>
-							builtImages[d.serviceName] ?? {
-								serviceName: d.serviceName,
-								name: typeof d.image === 'string' ? d.image : d.image.tag,
-								logs: 'Build skipped; image for service already exists.',
-								props: {},
-							},
-					),
-				)
-				.then(function(images) {
-					if (opts.app.application_type?.[0]?.is_legacy) {
-						const { deployLegacy } = require('../utils/deploy-legacy');
-
-						const msg = getChalk().yellow(
-							'Target application requires legacy deploy method.',
+					.filter(d => !!d)
+					.then(function(servicesToSkip) {
+						// multibuild takes in a composition and always attempts to
+						// build or pull all services. we workaround that here by
+						// passing a modified composition.
+						const compositionToBuild = _.cloneDeep(project.composition);
+						compositionToBuild.services = _.omit(
+							compositionToBuild.services,
+							servicesToSkip,
 						);
-						logger.logWarn(msg);
+						if (_.size(compositionToBuild.services) === 0) {
+							logger.logInfo(
+								'Everything is up to date (use --build to force a rebuild)',
+							);
+							return {};
+						}
+						return compose
+							.buildProject(
+								docker,
+								logger,
+								project.path,
+								project.name,
+								compositionToBuild,
+								opts.app.arch,
+								opts.app.device_type,
+								opts.buildEmulated,
+								opts.buildOpts,
+								composeOpts.inlineLogs,
+								opts.convertEol,
+								composeOpts.dockerfilePath,
+								composeOpts.nogitignore,
+							)
+							.then(builtImages => _.keyBy(builtImages, 'serviceName'));
+					})
+					.then(builtImages =>
+						project.descriptors.map(
+							d =>
+								builtImages[d.serviceName] ?? {
+									serviceName: d.serviceName,
+									name: typeof d.image === 'string' ? d.image : d.image.tag,
+									logs: 'Build skipped; image for service already exists.',
+									props: {},
+								},
+						),
+					)
+					// @ts-ignore slightly different return types of partial vs non-partial release
```

Comment:
This comment triggered a reformat, the rest of the code is unchanged apart from formatting